sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/db/pending/PendingTransactionEntity.kt

```diff
@@ -0,0 +1,142 @@
+package cash.z.ecc.android.sdk.internal.db.pending
+
+import androidx.room.ColumnInfo
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import cash.z.ecc.android.sdk.model.BlockHeight
+import cash.z.ecc.android.sdk.model.FirstClassByteArray
+import cash.z.ecc.android.sdk.model.PendingTransaction
+import cash.z.ecc.android.sdk.model.Zatoshi
+import cash.z.ecc.android.sdk.model.ZcashNetwork
+
+@Entity(tableName = "pending_transactions")
+data class PendingTransactionEntity(
+    @PrimaryKey(autoGenerate = true)
+    val id: Long = 0,
+    val toAddress: String,
+    val value: Long,
+    val memo: ByteArray?,
+    val accountIndex: Int,
+    val minedHeight: Long = NO_BLOCK_HEIGHT,
+    val expiryHeight: Long = NO_BLOCK_HEIGHT,
+
+    val cancelled: Int = 0,
+    val encodeAttempts: Int = -1,
+    val submitAttempts: Int = -1,
+    val errorMessage: String? = null,
+    val errorCode: Int? = null,
+    val createTime: Long = System.currentTimeMillis(),
+    @ColumnInfo(typeAffinity = ColumnInfo.BLOB)
+    val raw: ByteArray = byteArrayOf(),
+    @ColumnInfo(typeAffinity = ColumnInfo.BLOB)
+    val rawTransactionId: ByteArray? = byteArrayOf()
+) {
+    fun toPendingTransaction(zcashNetwork: ZcashNetwork) = PendingTransaction(
+        id = id,
+        value = Zatoshi(value),
+        memo = memo?.let { FirstClassByteArray(it) },
+        raw = FirstClassByteArray(raw),
+        toAddress = toAddress,
+        accountIndex = accountIndex,
+        minedHeight = if (minedHeight == NO_BLOCK_HEIGHT) {
+            null
+        } else {
+            BlockHeight.new(zcashNetwork, minedHeight)
+        },
+        expiryHeight = if (expiryHeight == NO_BLOCK_HEIGHT) {
+            null
+        } else {
+            BlockHeight.new(zcashNetwork, expiryHeight)
+        },
+        cancelled = cancelled,
+        encodeAttempts = encodeAttempts,
+        submitAttempts = submitAttempts,
+        errorMessage = errorMessage,
+        errorCode = errorCode,
+        createTime = createTime,
+        rawTransactionId = rawTransactionId?.let { FirstClassByteArray(it) }
+    )
+
+    @Suppress("ComplexMethod")
+    override fun equals(other: Any?): Boolean {
```

Comment:
Good question.

This is unfortunately correct as-is.

Carried over as JVM behavior, `ByteArray` and other arrays do referential equality instead of value equality.  That's why I've introduced `FirstClassByteArray` which helps make the wrapped byte arrays behave more like "first class" objects that obey equals, hashcode, and so on.

Because PendingTransactionEntity is used by Room, we're stuck with ByteArray as a field type and cannot use FirstClassByteArray here.